src/hotspot/share/code/nmethod.cpp

```diff
@@ -2122,7 +2122,7 @@ void nmethod::unlink() {
   // Clear the link between this nmethod and a HotSpotNmethod mirror
   JVMCINMethodData* nmethod_data = jvmci_nmethod_data();
   if (nmethod_data != nullptr) {
-    nmethod_data->invalidate_nmethod_mirror(this);
+    nmethod_data->invalidate_nmethod_mirror(this, is_cold() ? nmethod::InvalidationReason::GC_UNLINKING_COLD : nmethod::InvalidationReason::GC_UNLINKING);
```

Comment:
So then this is the heart of what you're after?  Maybe `GC_UNLINKING_COLD` should be `COLD_UNLOADING`?  GC is doing it but it's not for GC reasons.  `GC_UNLINKING` might better be `GC_UNLOADING`.  I don't think the `nmethod::unlink` name is a good one to propagate into enum names.

`is_cold` isn't completely reliable here as an method could both have dead oops and be cold.  I guess that probably doesn't matter though.

Reply:
I'll do the renaming, good point there.

> is_cold isn't completely reliable here as an method could both have dead oops and be cold. I guess that probably doesn't matter though.

I don't think that matter because since the nmethod is cold right now it could have been eliminated for that reason at any moment.